### PR TITLE
Automatically generate API References

### PR DESCRIPTION
**Pull request by linshokaku**
Apply autosummary recursively to automatically generate all API documentation.

cf. #723

### CODE REVIEW

File: docs/source/reference/index.rst [hunk 1]
```diff
@@ -1,147 +1,143 @@
-.. module:: pytorch_pfn_extras
```
Comment (linshokaku) [inline docs/source/reference/index.rst:1]: Remove the module directive, as it impacts the links to the pytorch_pfn_extras package

File: docs/source/templates/autosummary/module.rst [hunk 2]
```diff
@@ -0,0 +1,68 @@
+{{ fullname | escape | underline}}
```
Comment (linshokaku) [inline docs/source/templates/autosummary/module.rst:1]: clone from https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/autosummary/templates/autosummary/module.rst and modified.

File: docs/requirements.txt [hunk 3]
```diff
@@ -1,6 +1,6 @@
-sphinx==4.1.0
-pydata_sphinx_theme==0.6.3
-myst-parser==0.15.1
+sphinx==7.1.0
+pydata_sphinx_theme==0.13.3
```
Comment (kmaehashi) [inline docs/requirements.txt:2]: There is a known issue with RTD in the latest `pydata_sphinx_theme` so let's stick with the latest working one.
```suggestion
pydata_sphinx_theme==0.11.0
```

c.f. https://github.com/cupy/cupy/pull/7375

File: docs/source/conf.py [hunk 4]
```diff
@@ -36,7 +36,7 @@
 autodoc_typehints = "description"
 
 # Add any paths that contain templates here, relative to this directory.
-templates_path = ["_templates"]
+templates_path = ["templates"]
```
Comment (kmaehashi) [inline docs/source/conf.py:39]: Question: Is this rename needed?

File: docs/source/reference/index.rst [hunk 5]
```diff
@@ -1,147 +1,143 @@
...
 
-   engine.create_trainer
-   engine.create_evaluator
-   handler.BaseLogic
-   handler.Logic
-   handler.BaseHandler
-   handler.Handler
-   runtime.BaseRuntime
-   runtime.PyTorchRuntime
+   pytorch_pfn_extras.engine.create_trainer
```
Comment (kmaehashi) [inline docs/source/reference/index.rst:23]: Question: Is it possible to do this without adding `pytorch_pfn_extras.` prefix?
The columns seems too narrow... 😓
https://pytorch-pfn-extras--731.org.readthedocs.build/en/731/reference/index.html#extensions

**Review by kmaehashi — CHANGES_REQUESTED** [review]
Sorry for the delay :bow:

File: docs/requirements.txt (hunk 3 above)
Comment (linshokaku) [reply docs/requirements.txt]: Fixed it!

File: docs/source/conf.py (hunk 4 above)
Comment (linshokaku) [reply docs/source/conf.py]: There's no need. Use "_templates".

File: docs/source/reference/index.rst (hunk 5 above)
Comment (linshokaku) [reply docs/source/reference/index.rst]: By using currentmodule, it has been identified that it's possible to change the module within a file. Therefore, this will be utilized to remove the prefix.

**Comment by kmaehashi** [comment]
Thanks, LGTM!

**Comment by kmaehashi** [comment]
Thanks, LGTM!
